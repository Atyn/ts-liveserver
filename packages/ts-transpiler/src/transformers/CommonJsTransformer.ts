import TypeScript, { TypeChecker } from 'typescript'

/*
Transpile CommonJS to ES6 module
*/
export default class CommonJsTransformer
	implements TypeScript.CustomTransformer {
	private context: TypeScript.TransformationContext
	private counter = 0
	constructor(context: TypeScript.TransformationContext) {
		this.context = context
	}
	transformBundle(): TypeScript.Bundle {
		throw new Error('Method not implemented.')
	}
	public transformSourceFile(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const requireInTopScope = this.requireTopScope(sourceFile)
		const withoutModule = this.stripModule(requireInTopScope)
		const withoutDefineProperty = this.convertDefinePropery(withoutModule)
		const esmExport = this.convertToEsmExport(withoutDefineProperty)
		return this.convertToEsmImport(esmExport)
	}
	// Generate a file unique variable name
	private generateUniqueName() {
		this.counter++
		return 'GENERATED_VAR_BY_TRANSFORMER_' + String(this.counter)
	}
	// Convert all Object.defineProperty(exports, "hello", Hello);"
	private convertDefinePropery(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (node: TypeScript.Node): TypeScript.Node => {
			if (
				TypeScript.isExpressionStatement(node) &&
				TypeScript.isCallExpression(node.expression) &&
				TypeScript.isPropertyAccessExpression(node.expression.expression) &&
				TypeScript.isIdentifier(node.expression.expression.expression) &&
				TypeScript.isIdentifier(node.expression.expression.name) &&
				node.expression.expression.expression.getText() === 'Object' &&
				node.expression.expression.name.getText() === 'defineProperty' &&
				TypeScript.isIdentifier(node.expression.arguments[0]) &&
				node.expression.arguments[0].getText() === 'exports'
			) {
				const secondArgument = node.expression.arguments[1]
				const thirdArgument = node.expression.arguments[2]
				if (
					TypeScript.isStringLiteral(secondArgument) &&
					TypeScript.isIdentifier(thirdArgument)
				) {
					return TypeScript.factory.createExportDeclaration(
						undefined,
						undefined,
						false,
						TypeScript.factory.createNamedExports([
							TypeScript.factory.createExportSpecifier(
								thirdArgument,
								secondArgument.text,
							),
						]),
					)
				}
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		return TypeScript.visitNode(sourceFile, visit)
	}
	// Top level CommonJS to ESM
	private convertToEsmImport(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (
			node: TypeScript.Node,
		): TypeScript.Node | TypeScript.Node[] => {
			// Import without name
			if (
				TypeScript.isExpressionStatement(node) &&
				TypeScript.isCallExpression(node.expression) &&
				TypeScript.isIdentifier(node.expression.expression) &&
				node.expression.expression.getText() === 'require' &&
				node.expression.arguments.length === 1
			) {
				const argument = node.expression.arguments[0]
				if (TypeScript.isStringLiteralLike(argument)) {
					return TypeScript.factory.createImportDeclaration(
						undefined,
						undefined,
						undefined,
						TypeScript.factory.createStringLiteral(argument.text),
					)
				}
			}
			// Import with reference
			if (
				TypeScript.isVariableStatement(node) &&
				TypeScript.isVariableDeclarationList(node.declarationList)
			) {
				const importDeclarations: TypeScript.ImportDeclaration[] = []
				const variableDeclarations: TypeScript.VariableDeclaration[] = []
				for (const variableDeclaration of node.declarationList.declarations) {
					if (
						variableDeclaration.initializer &&
						TypeScript.isCallExpression(variableDeclaration.initializer) &&
						TypeScript.isIdentifier(
							variableDeclaration.initializer.expression,
						) &&
						variableDeclaration.initializer.expression.getText() ===
							'require' &&
						variableDeclaration.initializer.arguments.length === 1
					) {
						const argument = variableDeclaration.initializer.arguments[0]
						if (
							TypeScript.isStringLiteral(argument) &&
							TypeScript.isIdentifier(variableDeclaration.name)
						) {
							importDeclarations.push(
								TypeScript.factory.createImportDeclaration(
									undefined,
									undefined,
									TypeScript.factory.createImportClause(
										false,
										undefined,
										TypeScript.factory.createNamespaceImport(
											variableDeclaration.name,
										),
									),
									argument,
								),
							)
						} else if (
							TypeScript.isStringLiteral(argument) &&
							TypeScript.isObjectBindingPattern(variableDeclaration.name)
						) {
							importDeclarations.push(
								TypeScript.factory.createImportDeclaration(
									undefined,
									undefined,
									TypeScript.factory.createImportClause(
										false,
										undefined,
										TypeScript.factory.createNamedImports(
											variableDeclaration.name.elements.map((bindingElement) =>
												TypeScript.factory.createImportSpecifier(
													bindingElement.propertyName as TypeScript.Identifier,
													bindingElement.name as TypeScript.Identifier,
												),
											),
										),
									),
									TypeScript.factory.createStringLiteral(argument.text),
								),
							)
						}
					} else {
						variableDeclarations.push(variableDeclaration)
					}
				}
				if (variableDeclarations.length === 0) {
					return importDeclarations
				} else {
					return [
						...importDeclarations,
						TypeScript.factory.updateVariableStatement(
							node,
							undefined,
							TypeScript.factory.updateVariableDeclarationList(
								node.declarationList,
								variableDeclarations,
							),
						),
					]
				}
			}
			return node
		}
		return TypeScript.visitEachChild(sourceFile, visit, this.context)
	}
	// Top level CommonJS to ESM
	private convertToEsmExport(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (
			node: TypeScript.Node,
		): TypeScript.Node | TypeScript.Node[] => {
			if (
				TypeScript.isExpressionStatement(node) &&
				TypeScript.isBinaryExpression(node.expression) &&
				node.expression.operatorToken.kind === TypeScript.SyntaxKind.EqualsToken
			) {
				// exports.something = something;
				if (
					TypeScript.isPropertyAccessExpression(node.expression.left) &&
					TypeScript.isIdentifier(node.expression.left.name) &&
					TypeScript.isIdentifier(node.expression.left.expression) &&
					node.expression.left.expression.getText() === 'exports'
				) {
					// exports.default = something;
					if (node.expression.left.name.getText() === 'default') {
						return TypeScript.factory.createExportAssignment(
							undefined,
							undefined,
							undefined,
							node.expression.right,
						)
					}
					// exports.something = anIdentifier;
					else if (TypeScript.isIdentifier(node.expression.right)) {
						return TypeScript.factory.createExportDeclaration(
							undefined,
							undefined,
							false,
							TypeScript.factory.createNamedExports([
								TypeScript.factory.createExportSpecifier(
									node.expression.right,
									node.expression.left.name,
								),
							]),
						)
					} else {
						const newIdentifierName = this.generateUniqueName()
						const variableStatement = TypeScript.factory.createVariableStatement(
							undefined,
							[
								TypeScript.factory.createVariableDeclaration(
									newIdentifierName,
									undefined,
									undefined,
									node.expression.right,
								),
							],
						)
						return [
							variableStatement,
							TypeScript.factory.createExportDeclaration(
								undefined,
								undefined,
								false,
								TypeScript.factory.createNamedExports([
									TypeScript.factory.createExportSpecifier(
										TypeScript.factory.createIdentifier(newIdentifierName),
										node.expression.left.name,
									),
								]),
							),
						]
					}
				}
				// exports = something;
				if (
					TypeScript.isIdentifier(node.expression.left) &&
					node.expression.left.getText() === 'exports'
				) {
					// exports = { a: false, b: true }
					if (TypeScript.isObjectLiteralExpression(node.expression.right)) {
						const exportSpecifiers: TypeScript.ExportSpecifier[] = []
						for (const property of node.expression.right.properties) {
							if (
								TypeScript.isPropertyAssignment(property) &&
								TypeScript.isIdentifier(property.name) &&
								TypeScript.isIdentifier(property.initializer)
							) {
								exportSpecifiers.push(
									TypeScript.factory.createExportSpecifier(
										property.initializer,
										property.name,
									),
								)
							}
						}
						return TypeScript.factory.createExportDeclaration(
							undefined,
							undefined,
							false,
							TypeScript.factory.createNamedExports(exportSpecifiers),
						)
					}
					// exports = require('hello.js');
					else if (
						TypeScript.isCallExpression(node.expression.right) &&
						node.expression.right.expression.getText() === 'require' &&
						TypeScript.isStringLiteral(node.expression.right.arguments[0])
					) {
						return TypeScript.factory.createExportDeclaration(
							undefined,
							undefined,
							false,
							undefined,
							node.expression.right.arguments[0],
						)
					}
					// exports = 'hello' or exports = Hello
					else {
						return TypeScript.factory.createExportAssignment(
							undefined,
							undefined,
							undefined,
							node.expression.right,
						)
					}
				}
			}
			return node
		}
		return TypeScript.visitEachChild(sourceFile, visit, this.context)
	}
	// module.exports -> exports
	private stripModule(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const visit = (node: TypeScript.Node): TypeScript.Node => {
			if (
				TypeScript.isPropertyAccessExpression(node) &&
				TypeScript.isIdentifier(node.expression) &&
				TypeScript.isIdentifier(node.name) &&
				node.expression.getText() === 'module' &&
				node.name.getText() === 'exports'
			) {
				return node.name
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		return TypeScript.visitNode(sourceFile, visit)
	}
	// Move all require-calls to top-scope
	private requireTopScope(
		sourceFile: TypeScript.SourceFile,
	): TypeScript.SourceFile {
		const newStatements: TypeScript.VariableStatement[] = []
		const visit = (node: TypeScript.Node): TypeScript.Node => {
			const inRootScope =
				node?.parent?.parent?.parent?.parent &&
				TypeScript.isVariableDeclaration(node.parent) &&
				TypeScript.isVariableDeclarationList(node.parent.parent) &&
				TypeScript.isVariableStatement(node.parent.parent.parent) &&
				TypeScript.isSourceFile(node.parent.parent.parent.parent)
			if (
				TypeScript.isCallExpression(node) &&
				TypeScript.isIdentifier(node.expression) &&
				node.arguments.length === 1 &&
				node.expression.getText() === 'require' &&
				inRootScope === false
			) {
				const newIdentifierName = this.generateUniqueName()
				newStatements.push(
					TypeScript.factory.createVariableStatement(undefined, [
						TypeScript.factory.createVariableDeclaration(
							newIdentifierName,
							undefined,
							undefined,
							node,
						),
					]),
				)
				return TypeScript.factory.createIdentifier(newIdentifierName)
			}
			return TypeScript.visitEachChild(node, visit, this.context)
		}
		const changedSourceFile = TypeScript.visitNode(sourceFile, visit)
		return TypeScript.factory.updateSourceFile(changedSourceFile, [
			...newStatements,
			...changedSourceFile.statements,
		])
	}
}
