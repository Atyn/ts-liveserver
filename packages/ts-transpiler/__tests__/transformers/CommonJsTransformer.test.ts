import CommonJsTransformer from '../../src/transformers/CommonJsTransformer'
import TypeScript from 'typescript'

describe('CommonJsTransformer', () => {
	it('Should convert to default export to ES6', async () => {
		const input = 'module.exports = {};'
		const output = 'export default {};'
		expect(await transform(input)).toBe(output)
	})
	it('Should convert to default export to ES6', async () => {
		const input = 'module.exports = {};'
		const output = 'export default {};'
		expect(await transform(input)).toBe(output)
	})
	it('Should convert default import to ES6', async () => {
		const input = 'const Hello = require("./hello.ts");'
		const output = 'import Hello from "./hello.ts";'
		expect(await transform(input)).toBe(output)
	})
	it('Should convert simple import', async () => {
		const input = 'require("./hello.ts");'
		const output = 'import "./hello.ts";'
		expect(await transform(input)).toBe(output)
	})
	it('Should convert named import', async () => {
		const input = 'const { Hello } = require("./hello.ts");'
		const output = 'import { Hello } from "./hello.ts";'
		expect(await transform(input)).toBe(output)
	})
	it('Should convert to default export to ES6', async () => {
		const input = 'module.exports.hello = Hello;'
		const output = 'export { Hello as hello };'
		expect(await transform(input)).toBe(output)
	})
	it('Should convert exports.name', async () => {
		const input = 'exports.hello = Hello;'
		const output = 'export { Hello as hello };'
		expect(await transform(input)).toBe(output)
	})
	it('Should convert exports.default', async () => {
		const input = 'exports.default = Hello;'
		const output = 'export default Hello;'
		expect(await transform(input)).toBe(output)
	})
})

const compilerOptions: TypeScript.CompilerOptions = {
	allowJs: true,
	jsx: TypeScript.JsxEmit.React,
	checkJs: false,
	noResolve: false,
	esModuleInterop: true,
	skipLibCheck: true,
	pretty: true,
	allowUnreachableCode: true,
	target: TypeScript.ScriptTarget.ES2020,
	declaration: false,
	module: TypeScript.ModuleKind.ES2020,
	moduleResolution: TypeScript.ModuleResolutionKind.NodeJs,
	sourceMap: false,
}

const transformers: TypeScript.CustomTransformers = {
	after: [(context) => new CommonJsTransformer(context)],
}

async function transform(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
		transformers: transformers,
	})
	return results.outputText.trim()
}
