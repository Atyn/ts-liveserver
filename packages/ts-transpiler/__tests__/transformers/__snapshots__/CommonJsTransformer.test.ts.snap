// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonJsTransformer Exports Should convert ObjectLiteralExpression 1`] = `
"let exports = {}, module = { exports: exports };
exports = { a: b, c: d };
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.a, GENERATED_VAR_BY_TRANSFORMER_2 = exports.c;
export { GENERATED_VAR_BY_TRANSFORMER_1 as a, GENERATED_VAR_BY_TRANSFORMER_2 as c };
export default exports;"
`;

exports[`CommonJsTransformer Exports Should convert define.property for identifier 1`] = `
"let exports = {}, module = { exports: exports };
exports.a = b;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.a;
export { GENERATED_VAR_BY_TRANSFORMER_1 as a };
export default exports;"
`;

exports[`CommonJsTransformer Exports Should convert define.property with getter 1`] = `
"let exports = {}, module = { exports: exports };
exports.hello = parts_js_1.PropertyPart;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.hello;
export { GENERATED_VAR_BY_TRANSFORMER_1 as hello };
export default exports;"
`;

exports[`CommonJsTransformer Exports Should convert define.property with value 1`] = `
"let exports = {}, module = { exports: exports };
exports.__esModule = true;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.__esModule;
export { GENERATED_VAR_BY_TRANSFORMER_1 as __esModule };
export default exports;"
`;

exports[`CommonJsTransformer Exports Should convert exports.default 1`] = `
"let exports = {}, module = { exports: exports };
exports.default = Hello;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.default;
export { GENERATED_VAR_BY_TRANSFORMER_1 as default };"
`;

exports[`CommonJsTransformer Exports Should convert exports.name 1`] = `
"let exports = {}, module = { exports: exports };
exports.hello = Hello;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.hello;
export { GENERATED_VAR_BY_TRANSFORMER_1 as hello };
export default exports;"
`;

exports[`CommonJsTransformer Exports Should convert exports.name with same name as the parameter 1`] = `
"let exports = {}, module = { exports: exports };
const hello = null;
exports.hello = hello;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.hello;
export { GENERATED_VAR_BY_TRANSFORMER_1 as hello };
export default exports;"
`;

exports[`CommonJsTransformer Exports Should convert named export 1`] = `
"let exports = {}, module = { exports: exports };
exports.hello = Hello;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.hello;
export { GENERATED_VAR_BY_TRANSFORMER_1 as hello };
export default exports;"
`;

exports[`CommonJsTransformer Exports Should convert to default export to ES6 1`] = `
"let exports = {}, module = { exports: exports };
exports = Hello;
export default exports;"
`;

exports[`CommonJsTransformer Exports Should export nested expression 1`] = `
"let exports = {}, module = { exports: exports };
(exports = function (key) {
    return key;
})('versions', []).push({
    version: '3.6.5',
});
export default exports;"
`;

exports[`CommonJsTransformer Exports Should export one default 1`] = `
"let exports = {}, module = { exports: exports };
exports.default = \\"hello\\";
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.default;
export { GENERATED_VAR_BY_TRANSFORMER_1 as default };"
`;

exports[`CommonJsTransformer Exports Should export one default 2`] = `
"let exports = {}, module = { exports: exports };
export default class Hello {
}
;
exports.hello = \\"hello\\";
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.hello;
export { GENERATED_VAR_BY_TRANSFORMER_1 as hello };"
`;

exports[`CommonJsTransformer Exports Should move exports to top scope 1`] = `
"let exports = {}, module = { exports: exports };
if (typeof define === 'function' && define.amd) {
    define(function () {
        return LZString;
    });
}
else if (typeof module !== 'undefined' && module != null) {
    exports = LZString;
}
export default exports;"
`;

exports[`CommonJsTransformer Exports Should not override current export 1`] = `
"let exports = {}, module = { exports: exports };
exports.a = \\"a\\";
exports.b = \\"b\\";
export { hello as a };
export default \\"yoyo\\";
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.b;
export { GENERATED_VAR_BY_TRANSFORMER_1 as b };"
`;

exports[`CommonJsTransformer Forwards Should convert redirects 1`] = `
"let exports = {}, module = { exports: exports };
exports = require(\\"./hello.js\\");
export * from \\"./hello.js\\";
export default exports;"
`;

exports[`CommonJsTransformer Forwards Should convert redirects 2`] = `
"let exports = {}, module = { exports: exports };
import WrenchIcon from './icons/WrenchIcon.js';
exports = { WrenchIcon };
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.WrenchIcon;
export { GENERATED_VAR_BY_TRANSFORMER_1 as WrenchIcon };
export default exports;"
`;

exports[`CommonJsTransformer Import Should convert default import to ES6 1`] = `
"let exports = {}, module = { exports: exports };
import Hello from \\"./hello.ts\\";
export default exports;"
`;

exports[`CommonJsTransformer Import Should convert named import 1`] = `
"let exports = {}, module = { exports: exports };
import { Hello } from \\"./hello.ts\\";
export default exports;"
`;

exports[`CommonJsTransformer Import Should convert require in sub-scope 1`] = `
"let exports = {}, module = { exports: exports };
import GENERATED_VAR_BY_TRANSFORMER_1 from \\"hello.js\\";
{
    const hello = GENERATED_VAR_BY_TRANSFORMER_1;
}
export default exports;"
`;

exports[`CommonJsTransformer Import Should convert require without reference in sub-scope 1`] = `
"let exports = {}, module = { exports: exports };
import \\"hello.js\\";
if (true) {
    void 0;
}
export default exports;"
`;

exports[`CommonJsTransformer Import Should convert simple import 1`] = `
"let exports = {}, module = { exports: exports };
import \\"./hello.ts\\";
export default exports;"
`;

exports[`CommonJsTransformer Should read from internal export Should convert to multiple commands 1`] = `
"let exports = {}, module = { exports: exports };
exports.lastAttributeNameRegex = exports.createMarker = exports.isTemplatePartActive = void 0;
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.lastAttributeNameRegex, GENERATED_VAR_BY_TRANSFORMER_2 = exports.createMarker, GENERATED_VAR_BY_TRANSFORMER_3 = exports.isTemplatePartActive;
export { GENERATED_VAR_BY_TRANSFORMER_1 as lastAttributeNameRegex, GENERATED_VAR_BY_TRANSFORMER_2 as createMarker, GENERATED_VAR_BY_TRANSFORMER_3 as isTemplatePartActive };
export default exports;"
`;

exports[`CommonJsTransformer Should read from internal export Should red from internal export "hello" 1`] = `
"let exports = {}, module = { exports: exports };
exports.hello = \\"Hello\\";
console.log(exports.hello);
const GENERATED_VAR_BY_TRANSFORMER_1 = exports.hello;
export { GENERATED_VAR_BY_TRANSFORMER_1 as hello };
export default exports;"
`;
